cmd/postgres_exporter/postgres_exporter.go

```diff
@@ -1215,22 +1215,34 @@ func newDesc(subsystem, name, help string, labels prometheus.Labels) *prometheus
 }
 
 func queryDatabases(server *Server) ([]string, error) {
-	rows, err := server.db.Query("SELECT datname FROM pg_database WHERE datallowconn = true AND datistemplate = false AND datname != current_database()") // nolint: safesql
+	query := `SELECT datname 
+                FROM pg_database  
+			   WHERE datallowconn = true 
+			     AND datistemplate = false 
+				 AND has_database_privilege(current_user, datname, 'connect')`
+
+	rows, err := server.db.Query(query)
 	if err != nil {
 		return nil, fmt.Errorf("error retrieving databases: %v", err)
 	}
 	defer rows.Close() // nolint: errcheck
 
 	var databaseName string
 	result := make([]string, 0)
+
 	for rows.Next() {
 		err = rows.Scan(&databaseName)
 		if err != nil {
 			return nil, errors.New(fmt.Sprintln("error retrieving rows:", err))
 		}
+
 		result = append(result, databaseName)
 	}
 
+	if rows.Err() != nil {
+		return nil, errors.New(fmt.Sprintln("error retrieving rows:", err))
```

Comment:
May be `fmt.Errorf` ?

Reply:
I didn't want to change it because upstream is not using newer libraries. 
The correct way would be use `errors.Wrap`. There are other parts where the exporting is comparing error types instead using `errors.As`, etc, but to not to differ too much from upstream, I didn't want to change them.